cedar-policy-core/src/validator/typecheck.rs

```diff
@@ -1929,17 +1941,86 @@ impl<'a> SingleEnvTypechecker<'a> {
                             rhs_expr,
                         ),
 
-                    // If none of the cases apply, then all we know is that `in` has
-                    // type boolean. Importantly for partial schema
-                    // validation, this case captures an `in` between entity
-                    // literals where the LHS is not an action defined in
-                    // the schema and does not have an entity type defined
-                    // in the schema.
-                    _ => TypecheckAnswer::success(
-                        ExprBuilder::with_data(Some(Type::primitive_boolean()))
-                            .with_same_source_loc(in_expr)
-                            .is_in(lhs_expr, rhs_expr),
-                    ),
+                    _ => match (lhs_expr.data(), rhs_expr.data()) {
+                        // If both sides have action entity types, we conservatively
+                        // check if any of the actions of type `rhs_ety` have descendant
+                        // action entities with type `lhs_ety`. If there is none, we
+                        // can soundly type it as false.
+                        (
+                            Some(Type::EntityOrRecord(EntityRecordKind::ActionEntity {
+                                name: lhs_ety,
+                                ..
+                            })),
+                            Some(Type::EntityOrRecord(EntityRecordKind::ActionEntity {
+                                name: rhs_ety,
+                                ..
+                            })),
+                        ) if !self.check_action_descendant(lhs_ety, rhs_ety) => {
+                            TypecheckAnswer::success(
+                                ExprBuilder::with_data(Some(Type::False))
+                                    .with_same_source_loc(in_expr)
+                                    .is_in(lhs_expr, rhs_expr),
+                            )
+                        }
+
+                        // Similar to the case above, but checks the case when the RHS
+                        // could be a collection of entity types.
+                        (
+                            Some(Type::EntityOrRecord(EntityRecordKind::ActionEntity {
+                                name: lhs_ety,
+                                ..
+                            })),
+                            Some(Type::EntityOrRecord(EntityRecordKind::Entity(rhs_etys))),
+                        ) if rhs_etys
+                            .iter()
+                            .all(|rhs_ety| !self.check_action_descendant(lhs_ety, rhs_ety)) =>
```

Comment:
is this missing a check for `EntityRecordKind::ActionEntity`?  In the single case, we check that it's `ActionEntity`, but in the multiple case, we do not, which seems inconsistent